cmd/kube-controller-manager/app/controllermanager_test.go

```diff
@@ -115,6 +121,56 @@ func TestNewControllerDescriptorsShouldNotPanic(t *testing.T) {
 	NewControllerDescriptors()
 }
 
+func TestGracefulControllers(t *testing.T) {
+	controllerDescriptorMap := NewControllerDescriptors()
+	gracefulControllers := sets.NewString(names.ClusterRoleAggregationController)
+
+	for controllerName, controllerDesc := range controllerDescriptorMap {
+		t.Run(controllerName, func(t *testing.T) {
+			defer goleak.VerifyNone(t)
+			var wg sync.WaitGroup
+			defer wg.Wait()
+
+			// Start with a single controller as test
+			if !gracefulControllers.Has(controllerName) {
+				t.Skipf("Controller %s is not in the graceful shutdown test list", controllerName)
+			}
+
+			ctx, cancel := context.WithCancel(context.Background())
+			defer cancel()
+			testClientset := fake.NewSimpleClientset()
+			testClientBuilder := TestClientBuilder{clientset: testClientset}
+			testInformerFactory := informers.NewSharedInformerFactoryWithOptions(testClientset, 3*time.Minute)
+			config, err := options.NewDefaultComponentConfig()
+			if err != nil {
+				t.Fatal(err)
+			}
+			controllerContext := ControllerContext{
+				ClientBuilder:                   testClientBuilder,
+				ComponentConfig:                 config,
+				InformerFactory:                 testInformerFactory,
+				ObjectOrMetadataInformerFactory: testInformerFactory,
+				InformersStarted:                make(chan struct{}),
+			}
+
+			// Cloud provider controllers should be skipped
+			if controllerDesc.IsCloudProviderController() {
+				return
+			}
+			c, err := controllerDesc.GetControllerConstructor()(ctx, controllerContext, controllerName)
+			if err != nil {
+				t.Fatal(err)
+			}
+			if c == nil {
+				return
+			}
+			wg.Go(func() {
+				c.Run(ctx)
+			})
```

Comment:
Shouldn't we sleep for some short time to allow `Run` to actually start anything?